File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

input = sys.stdin.readline

N, M, V = list(map(int, input().split()))

graph = [[False] * (N+1) for _ in range(N+1)]

for _ in range(M):
    r, c = list(map(int, input().split()))
    graph[r][c] = True
    graph[c][r] = True


def dfs():
    visited = [False] * (N+1)
    result = []

    def recursive(cur_node):

        visited[cur_node] = True
        result.append(cur_node)
        
        for next_node, isConnected in enumerate(graph[cur_node]):
            if isConnected and not visited[next_node]:
                recursive(next_node)


    recursive(V)
    
    for r in result:
        print(r, end=" ")


def bfs():
    result = [V]
    visited = [False] * (N+1)
    visited[V] = True

    q = deque()
    q.append(V)

    while q:
        cur_node = q.popleft()

        for next_node, isConnected in enumerate(graph[cur_node]):
            if isConnected and not visited[next_node]:
                result.append(next_node)
                q.append(next_node)
                visited[next_node] = True
    
    for r in result:
        print(r, end=" ")



dfs()
print()
bfs()

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.swift
import Foundation



func bfs(graph: [[Int]], v: Int) {
  var visited: [Bool] = Array(repeating: false, count: N+1)
  visited[v] = true
  var q: [Int] = [v]
  var result = "\(v) "
  
  while !q.isEmpty {
    let cur = q.removeFirst()
    
    for next in graph[cur] {
      if !visited[next] {
        visited[next] = true
        result += "\(next) "
        q.append(next)
      }
    }
  }
  
  print(result)
}

func dfs(graph: [[Int]], v: Int) {
  var visited: [Bool] = Array(repeating: false, count: N+1)
  var result = ""
  
  func recursive(v: Int) {
    visited[v] = true
    result += "\(v) "
    for next in graph[v] {
      if !visited[next] {
        recursive(v: next)
      }
    }
  }
  
  recursive(v: v)
  print(result)
}

let values = readLine()!.split(separator: " ").map { Int(String($0))! }
let N = values[0]
let M = values[1]
let V = values[2]

var graph: [[Int]] = Array(repeating: [], count: N+1)

for _ in 0..<M {
  let nodes = readLine()!.split(separator: " ").map { Int(String($0))! }
  let x = nodes[0]
  let y = nodes[1]
  
  graph[x].append(y)
  graph[y].append(x)
  
  graph[x].sort()
  graph[y].sort()
}

dfs(graph: graph, v: V)
bfs(graph: graph, v: V)
